flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/AdaptedRestartPipelinedRegionStrategyNG.java

```diff
@@ -0,0 +1,302 @@
+
+		for (ExecutionVertex ev : verticesToRestart) {
+			CoLocationGroup cgroup = ev.getJobVertex().getCoLocationGroup();
+			if (cgroup != null && !colGroups.contains(cgroup)){
+				cgroup.resetConstraints();
+				colGroups.add(cgroup);
+			}
+
+			ev.resetForNewExecution(restartTimestamp, globalModVersion);
+		}
+
+		// if there is checkpointed state, reload it into the executions
+		if (executionGraph.getCheckpointCoordinator() != null) {
+			// abort pending checkpoints to
+			// i) enable new checkpoint triggering without waiting for last checkpoint expired.
+			// ii) ensure the EXACTLY_ONCE semantics if needed.
+			executionGraph.getCheckpointCoordinator().abortPendingCheckpoints(
+				new CheckpointException(CheckpointFailureReason.JOB_FAILOVER_REGION));
+
+			final Map<JobVertexID, ExecutionJobVertex> involvedExecutionJobVertices =
+				getInvolvedExecutionJobVertices(verticesToRestart);
+			executionGraph.getCheckpointCoordinator().restoreLatestCheckpointedState(
+				involvedExecutionJobVertices, false, true);
+		}
+	}
+
+	private void rescheduleTasks(
+		final Set<ExecutionVertex> verticesToRestart,
+		final long globalModVersion) throws Exception {
+
+		final CompletableFuture<Void> newSchedulingFuture;
+		switch (executionGraph.getScheduleMode()) {
+
+			case LAZY_FROM_SOURCES:
+				newSchedulingFuture = AdaptedSchedulingUtils.scheduleLazy(verticesToRestart, executionGraph);
+				break;
+
+			case EAGER:
+				newSchedulingFuture = AdaptedSchedulingUtils.scheduleEager(verticesToRestart, executionGraph);
+				break;
+
+			default:
+				throw new JobException("Schedule mode is invalid.");
+		}
+
+		// if no global failover is triggered in the scheduling process,
+		// register a failure handling callback to the scheduling
+		if (isLocalFailoverValid(globalModVersion)) {
+			newSchedulingFuture.whenComplete(
+				(Void ignored, Throwable throwable) -> {
+					if (throwable != null) {
+						final Throwable strippedThrowable = ExceptionUtils.stripCompletionException(throwable);
+
+						if (!(strippedThrowable instanceof CancellationException)) {
+							// only fail if the scheduling future was not canceled
+							failGlobal(strippedThrowable);
+						}
+					}
+				});
+		}
+	}
+
+	private boolean isLocalFailoverValid(final long globalModVersion) {
+		// local failover is valid only if the job is still RUNNING and
+		// no global failover happens since the given globalModVersion is recorded
+		return executionGraph.getState() == JobStatus.RUNNING &&
+			executionGraph.getGlobalModVersion() == globalModVersion;
+	}
+
+	private CompletableFuture<?> cancelExecutionVertex(final ExecutionVertexID executionVertexId) {
+		return getExecutionVertex(executionVertexId).cancel();
+	}
+
+	private Map<JobVertexID, ExecutionJobVertex> getInvolvedExecutionJobVertices(
+		final Set<ExecutionVertex> executionVertices) {
+
+		Map<JobVertexID, ExecutionJobVertex> tasks = new HashMap<>(executionVertices.size());
```

Comment:
initial size for map is not necessary. see: https://stackoverflow.com/questions/564899/choosing-the-initial-capacity-of-a-hashset-with-an-expected-number-of-unique-val

Reply:
Ok.